### PR TITLE
New screws pickup env

### PR DESCRIPTION
**Pull request by NishanthJKumar**
A simple environment for which we need to sideline something other than `NextTo` in order to do well. This is inspired by a claw-game type situation where the gripper is so big that it picks up multiple objects at once.

Video of successful execution:

https://user-images.githubusercontent.com/12446934/168487539-29acbc92-d3f5-4679-947a-dae602f11cf4.mp4

### CODE REVIEW

**Comment by tomsilver** [comment]
love the rendering! can we fix this though?

image

File: src/envs/screws.py [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+from predicators.src.settings import CFG
+from predicators.src.structs import Action, Array, GroundAtom, Image, Object, \
+    ParameterizedOption, Predicate, State, Task, Type
+
+
+class ScrewsEnv(BaseEnv):
+    """Toy screw picking env.
+
+    In this environment, the agent controls a magnetic gripper that can
+    be used to pick up screws of different colors/sizes. The challennge
```
Comment (tomsilver) [inline src/envs/screws.py:20]: typo: challennge

File: src/envs/screws.py [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+    screw.
+    """
+    _gripper_width: ClassVar[float] = 0.5
+    _receptacle_width: ClassVar[float] = 0.6
+    _receptacle_height: ClassVar[float] = 0.2
+    _screw_width: ClassVar[float] = 0.05
+    _screw_height: ClassVar[float] = 0.05
+    _magnetic_field_dist: ClassVar[float] = 0.5
+    _num_screw_clusters: ClassVar[int] = 3
+    num_screws_train: ClassVar[List[int]] = CFG.screws_num_screws_train
```
Comment (tomsilver) [inline src/envs/screws.py:32]: tests sometimes get screwy if we do this. suggest instead setting these fields that use `CFG` in `__init__()`

Comment (tomsilver) [reply src/envs/screws.py:32]: (no pun intended)

File: src/envs/screws.py [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+    def __init__(self) -> None:
+        super().__init__()
+        # Types
+        self._screw_type = Type(
+            "screw", ["pose_x", "pose_y", "width", "height", "held"])
+        self._gripper_type = Type("gripper", ["pose_x", "pose_y", "width"])
+        self._receptacle_type = Type("receptacle",
+                                     ["pose_x", "pose_y", "width"])
+        # Predicates
+        self._ScrewPickupable = Predicate(
```
Comment (tomsilver) [inline src/envs/screws.py:49]: is this the same thing as the screw being on the ground? if so, rename to `ScrewOnGround`?

File: src/envs/screws.py [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+            "AboveReceptacle", [self._gripper_type, self._receptacle_type],
+            self._AboveReceptacle_holds)
+        self._HoldingScrew = Predicate("HoldingScrew", [self._screw_type],
+                                       self._HoldingScrew_holds)
+        self._ScrewInReceptacle = Predicate(
+            "ScrewInReceptacle", [self._screw_type, self._receptacle_type],
+            self._ScrewInReceptacle_holds)
+
+        # Options
+        self._MoveToScrew: ParameterizedOption = \
```
Comment (tomsilver) [inline src/envs/screws.py:62]: are these `: ParameterizedOption` annotations needed? (if not, could avoid the new line break too)

File: src/envs/screws.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+            utils.SingletonParameterizedOption(
+            # variables: [robot, screw to pick up]
+            # params: []
+            "MoveToScrew",
+            self.MoveToScrew_policy,
```
Comment (tomsilver) [inline src/envs/screws.py:67]: can all these policies be made private methods?

File: src/envs/screws.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+        self._DemagnetizeGripper: ParameterizedOption = \
+            utils.SingletonParameterizedOption(
+            # variables: [robot]
+            # params: []
+            "DemagnetizeGripper",
+            self.DemagnetizeGripper_policy,
+            types=[self._gripper_type])
+
+        # Static objects (always exist no matter the settings).
+        self.robot = Object("robby", self._gripper_type)
```
Comment (tomsilver) [inline src/envs/screws.py:92]: make private?

File: src/envs/screws.py [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+            receptacle_x - receptacle_width / 2.0,
+            receptacle_x + receptacle_width / 2.0
+        ], [receptacle_y, receptacle_y],
+                 color="green")
+
+        # Draw the gripper
+        gripper_x = state.get(self.robot, "pose_x")
+        gripper_y = state.get(self.robot, "pose_y")
+        gripper_width = state.get(self.robot, "width")
+        # NOTE: this 1.3 factor is just to make the gripper wide enough so that
```
Comment (tomsilver) [inline src/envs/screws.py:135]: the 1.3 thing here is liable to break if we later change the dimensions of stuff. can we accomplish the same thing by changing the conditions for picking so that the entirety of the screw is below the gripper (taking into account the widths of both)?

File: src/envs/screws.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+            [gripper_y - gripper_width / 4.0, gripper_y],
+            color="black")
+
+        # Find which object is the goal screw.
+        assert len(task.goal) == 1
+        goal_pred = list(task.goal)[0]
+        goal_screw = goal_pred.objects[0]
+        assert goal_screw.type.name == 'screw'
+
+        # Draw screws
```
Comment (tomsilver) [inline src/envs/screws.py:160]: nit: add period

File: src/envs/screws.py [hunk 9]
```diff
@@ -0,0 +1,509 @@
...
+            if screw == goal_screw:
+                color = 'gold'
+            else:
+                color = 'grey'
+            rect = plt.Rectangle((screw_x, screw_y),
+                                 screw_width,
+                                 screw_height,
+                                 edgecolor=color,
+                                 facecolor=color,
+                                 label=screw.name)
```
Comment (tomsilver) [inline src/envs/screws.py:175]: is this `label` argument used? it doesn't look like it

File: src/envs/screws.py [hunk 10]
```diff
@@ -0,0 +1,509 @@
...
+            ax.add_patch(rect)
+
+        img = utils.fig2data(fig, dpi=150)
```
Comment (tomsilver) [inline src/envs/screws.py:178]: remove `=150` since that's the default

File: src/envs/screws.py [hunk 11]
```diff
@@ -0,0 +1,509 @@
...
+        return {
+            self._MoveToScrew, self._MoveToReceptacle, self._MagnetizeGripper,
+            self._DemagnetizeGripper
+        }
+
+    @property
+    def action_space(self) -> Box:
+        # dimensions: [dx, dy, magnetized vs. unmagnetized].
+        lowers = np.array([
+            -(self.rz_x_ub - self.rz_x_lb), -(self.rz_y_ub - self.rz_y_lb), 0.0
```
Comment (tomsilver) [inline src/envs/screws.py:208]: suggest making variables like `rz_width = self.rz_x_ub - self.rz_x_lb` and reusing in `lowers` and `uppers`

File: src/envs/screws.py [hunk 12]
```diff
@@ -0,0 +1,509 @@
...
+        ],
+                          dtype=np.float32)
+        uppers = np.array([(self.rz_x_ub - self.rz_x_lb),
+                           (self.rz_y_ub - self.rz_y_lb), 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        # NOTE: currently, the agent can only pick things up or
```
Comment (tomsilver) [inline src/envs/screws.py:218]: why is this necessary? can't we always just process the move first, then process the magnetization?

File: src/envs/screws.py [hunk 13]
```diff
@@ -0,0 +1,509 @@
...
+
+        if magnetization > 0.5:
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
```
Comment (tomsilver) [inline src/envs/screws.py:233]: I don't care that much about this, but I'm wondering if we could get away without the notion of screw clusters, and just randomly sample positions of all the screws, period?

File: src/envs/screws.py [hunk 14]
```diff
@@ -0,0 +1,509 @@
...
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
+            for si in range(num_screws):
+                existing_xys = set(screw_name_to_pos.values())
+                while True:
+                    # sample a random iniitial position for the screw
```
Comment (tomsilver) [inline src/envs/screws.py:246]: capitalize here and elsewhere. also typo: iniitial

File: src/envs/screws.py [hunk 15]
```diff
@@ -0,0 +1,509 @@
...
+                    # such that it is on the ground, not in the
```
Comment (tomsilver) [inline src/envs/screws.py:247]: doesn't "on the ground" imply "not in the receptacle"?

File: src/envs/screws.py [hunk 16]
```diff
@@ -0,0 +1,509 @@
...
+        for _ in range(self._num_screw_clusters):
+            screw_cluster_xcenters.append(
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
+            for si in range(num_screws):
+                existing_xys = set(screw_name_to_pos.values())
```
Comment (tomsilver) [inline src/envs/screws.py:244]: maybe just keep track of the existing xs in this loop, and worry about the ys and the names after all of the xs have been determined?

File: src/envs/screws.py [hunk 17]
```diff
@@ -0,0 +1,509 @@
...
+                    y_pos = self.rz_y_lb + self._screw_height / 2.0
+                    screw_cluster_xcenter = rng.choice(screw_cluster_xcenters)
+                    x_pos = rng.normal(screw_cluster_xcenter)
+                    # if this position doesn't make this screw in collision
+                    # with another screw, then put the screw here.
+                    if self._surface_xy_is_valid(x_pos, y_pos, existing_xys):
+                        screw_name_to_pos[f"screw_{si}"] = (x_pos, y_pos)
+                        break
+            init_state, goal_atoms = \
+                self._get_init_state_and_goal_atoms_from_positions(
```
Comment (tomsilver) [inline src/envs/screws.py:258]: might as well just get the task itself?

File: src/envs/screws.py [hunk 18]
```diff
@@ -0,0 +1,509 @@
...
+            [(gripper_x + gripper_width / 2.0) + (self._screw_width * 1.3),
+             (gripper_x + gripper_width / 2.0) + (self._screw_width * 1.3)],
+            [gripper_y - gripper_width / 4.0, gripper_y],
+            color="black")
+
+        # Find which object is the goal screw.
+        assert len(task.goal) == 1
+        goal_pred = list(task.goal)[0]
+        goal_screw = goal_pred.objects[0]
+        assert goal_screw.type.name == 'screw'
```
Comment (tomsilver) [inline src/envs/screws.py:158]: can probably just remove this assertion, but if you want, you can do `goal_screw.is_instance(self._screw_type)` instead

File: src/envs/screws.py [hunk 19]
```diff
@@ -0,0 +1,509 @@
...
+            init_state, goal_atoms = \
+                self._get_init_state_and_goal_atoms_from_positions(
+                screw_name_to_pos, rng)
+            tasks.append(Task(init_state, goal_atoms))
+        return tasks
+
+    def _get_init_state_and_goal_atoms_from_positions(
+            self, screw_name_to_pos: Dict[str, Tuple[float, float]],
+            rng: np.random.Generator) -> Tuple[State, Set[GroundAtom]]:
+        data: Dict[Object, Array] = {}
```
Comment (tomsilver) [inline src/envs/screws.py:266]: I'm recently a fan of using `utils.create_state_from_dict` instead for this, for improved readibility. see stick button or doors for examples.

File: src/envs/screws.py [hunk 20]
```diff
@@ -0,0 +1,509 @@
...
+    rz_x_lb = -5.0
+    rz_x_ub = 5.0
+    rz_y_lb = 0.0
+    rz_y_ub = 5.0
+
+    def __init__(self) -> None:
+        super().__init__()
+        # Types
+        self._screw_type = Type(
+            "screw", ["pose_x", "pose_y", "width", "height", "held"])
```
Comment (tomsilver) [inline src/envs/screws.py:44]: if all screws are the same height and width, we probably don't need to include those in their state features

Comment (tomsilver) [reply src/envs/screws.py:44]: ditto for width in the next two types

File: src/envs/screws.py [hunk 21]
```diff
@@ -0,0 +1,509 @@
...
+        return State(data), goal_atoms
+
+    def _surface_xy_is_valid(self, x: float, y: float,
+                             existing_xys: Set[Tuple[float, float]]) -> bool:
+        # First, check that there isn't another screw at this location.
+        for (existing_x, existing_y) in existing_xys:
+            if (abs(existing_x - x) < self._screw_width
+                    and abs(existing_y - y) < self._screw_height):
+                return False
+        # Next, check that the location is valid.
```
Comment (tomsilver) [inline src/envs/screws.py:304]: shouldn't this part be guaranteed by the bounds with which you're sampling? if not, change those bounds

File: src/envs/screws.py [hunk 22]
```diff
@@ -0,0 +1,509 @@
...
+        y_is_valid = y == self.rz_y_lb + self._screw_height / 2.0
+        x_is_valid = (self.rz_x_lb + self._screw_width / 2.0) <= x <= (
+            self.rz_x_ub - self._screw_width / 2.0)
+        return x_is_valid and y_is_valid
+
+    def _transition_move(self, state: State, x: float, y: float) -> State:
```
Comment (tomsilver) [inline src/envs/screws.py:310]: these inputs should be `dx` and `dy`

File: src/envs/screws.py [hunk 23]
```diff
@@ -0,0 +1,509 @@
...
+        new_rx = rx + x
+        new_rx_min = new_rx - rwidth / 2.0
+        new_rx_max = new_rx + rwidth / 2.0
+        new_ry = ry + y
+
+        # If the new position is outside the boundaries, move till
+        # the robot collides with the boundary.
+        if new_rx_min < self.rz_x_lb:
+            new_rx = self.rz_x_lb + (rwidth / 2.0)
+        if new_rx_max > self.rz_x_ub:
```
Comment (tomsilver) [inline src/envs/screws.py:324]: elif

File: src/envs/screws.py [hunk 24]
```diff
@@ -0,0 +1,509 @@
...
+            new_rx = self.rz_x_ub - (rwidth / 2.0)
+        if new_ry < self.rz_y_lb:
+            new_ry = self.rz_y_lb
+        if new_ry > self.rz_y_ub:
```
Comment (tomsilver) [inline src/envs/screws.py:328]: elif

File: src/envs/screws.py [hunk 25]
```diff
@@ -0,0 +1,509 @@
...
+        self._receptacle_type = Type("receptacle",
+                                     ["pose_x", "pose_y", "width"])
+        # Predicates
+        self._ScrewPickupable = Predicate(
+            "ScrewPickupable", [self._gripper_type, self._screw_type],
+            self._ScrewPickupable_holds)
+        self._AboveReceptacle = Predicate(
+            "AboveReceptacle", [self._gripper_type, self._receptacle_type],
+            self._AboveReceptacle_holds)
+        self._HoldingScrew = Predicate("HoldingScrew", [self._screw_type],
```
Comment (tomsilver) [inline src/envs/screws.py:55]: i know there's only one robot, but if the other predicates have the robot as an argument, it makes sense for this one to also

File: src/envs/screws.py [hunk 26]
```diff
@@ -0,0 +1,509 @@
...
+        ry = state.get(self.robot, "pose_y")
+        all_screws = state.get_objects(self._screw_type)
+        for screw in all_screws:
+            if not self._HoldingScrew_holds(state, [screw]):
+                # Check if the gripper is in the right position to
+                # pick up this screw, and if so, move the screw to
+                # the gripper.
+                if self._ScrewPickupable_holds(state, [self.robot, screw]):
+                    screw_height = state.get(screw, "height")
+                    next_state.set(screw, "pose_y", ry - (screw_height / 2.0))
```
Comment (tomsilver) [inline src/envs/screws.py:360]: probably related to my rendering gripe, if `ry` is the center of the gripper, this calculation doesn't look quite right

File: src/envs/screws.py [hunk 27]
```diff
@@ -0,0 +1,509 @@
...
+        return next_state
+
+    def _transition_demagnetize(self, state: State) -> State:
+        next_state = state.copy()
+        all_screws = state.get_objects(self._screw_type)
+        for screw in all_screws:
+            if self._HoldingScrew_holds(state, [screw]):
+                # NOTE: This currently will drop all screws into the receptacle
+                # iff AboveReceptacle is true. Otherwise, it will drop all
+                # screws onto the floor. This neglects the case where the
```
Comment (tomsilver) [inline src/envs/screws.py:372]: hmm yeah not ideal. can we change the meaning of AboveReceptacle so that the gripper must be fully over it?

File: tests/envs/test_screws.py [hunk 28]
```diff
@@ -0,0 +1,64 @@
+"""Test cases for the cover environment."""
```
Comment (tomsilver) [inline tests/envs/test_screws.py:1]: screws environment

File: src/planning.py [hunk 29]
```diff
@@ -291,6 +291,7 @@ def _run_low_level_search(task: Task, option_model: _OptionModelBase,
     discovered_failures: List[Optional[_DiscoveredFailure]] = [
         None for _ in skeleton
     ]
+
```
Comment (tomsilver) [inline src/planning.py:294]: revert

**Review by tomsilver — COMMENTED** [review]
awesome work!! i'm really happy that we can learn operators for an environment like this :)

File: src/envs/screws.py (hunk 3 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:49]: no - this means that the gripper is hovering over the screw and close enough to pick it up. I originally called it `GripperAboveScrew`, but the problem is that it's not just that, but also that the gripper is close enough to magnetize. Maybe I'll call it `GripperCanPickScrew`?

File: src/envs/screws.py (hunk 5 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:67]: so the problem is that I call the `MoveToScrew_policy` in a test. Any thoughts on a good way to work around this?

File: src/envs/screws.py (hunk 6 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:92]: also need in a test...

File: src/envs/screws.py (hunk 13 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:233]: i did that initially, and it led to things generally being grouped further apart. I like the notion of clusters simply because it makes it more likely that we're going to see a lot of different number of screws being held when collecting training data.

File: src/envs/screws.py (hunk 12 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:218]: good point; this note is deprecated...

File: src/envs/screws.py (hunk 27 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:372]: i think it already does this, so this comment isn't actually accurate

File: src/envs/screws.py (hunk 26 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:360]: wait why? I feel like this is exactly right (`ry` is the center of the gripper, and `"pose_y"` is the center of the block, so we want to set `"pose_y"` to `ry - half_of_screw_height`)

File: src/envs/screws.py (hunk 21 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:304]: since we're using a normal distribution to sample, we can't enforce bounds

File: src/envs/screws.py (hunk 16 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:244]: i like the current structure; i think doing the y's separately looks clunky, and this way, we could easily extend to put things in different y positions as well in the future if necessary!

File: src/envs/screws.py (hunk 3 above)
Comment (tomsilver) [reply src/envs/screws.py]: got it, yeah `GripperCanPickScrew` sounds good

File: src/envs/screws.py (hunk 5 above)
Comment (tomsilver) [reply src/envs/screws.py:68]: what we usually do is get the option instead from `env.options` and use the policy through the option, like here: https://github.com/Learning-and-Intelligent-Systems/predicators/blob/master/tests/envs/test_doors_env.py#L188

but if you need the policy specifically, you can create a subclass of the main environment that exposes things, like here: https://github.com/Learning-and-Intelligent-Systems/predicators/blob/master/tests/envs/test_pybullet_blocks.py#L14

File: src/envs/screws.py (hunk 6 above)
Comment (tomsilver) [reply src/envs/screws.py:90]: you can get the robot from a state, like here: https://github.com/Learning-and-Intelligent-Systems/predicators/blob/master/tests/envs/test_doors_env.py#L55

File: src/envs/screws.py (hunk 13 above)
Comment (tomsilver) [reply src/envs/screws.py:213]: okay sounds good!

File: src/envs/screws.py (hunk 21 above)
Comment (tomsilver) [reply src/envs/screws.py:290]: ah right. FWIW, you could use a truncated normal, but this is fine too.

File: src/envs/screws.py (hunk 26 above)
Comment (tomsilver) [reply src/envs/screws.py]: don't you need to take into account the height of the gripper?

Comment (NishanthJKumar) [reply src/envs/screws.py]: The gripper is a flat line and so doesn't have a height - I think during rendering it's given a height because of the line thickness

Comment (tomsilver) [reply src/envs/screws.py]: ah, got it. in that case, you can just change the rendering. probably easier to make it a rectangle

**Comment by tomsilver** [comment]
I notice when running oracle that the first skeleton often doesn't succeed, and it needs 2 skeletons to work; is that expected? @NishanthJKumar

File: src/envs/screws.py [hunk 30]
```diff
@@ -0,0 +1,482 @@
...
+        lowers = np.array([-self._rz_width, -self._rz_height, 0.0],
+                          dtype=np.float32)
+        uppers = np.array([self._rz_width, self._rz_height, 1.0],
+                          dtype=np.float32)
+        return Box(lowers, uppers)
+
+    def simulate(self, state: State, action: Action) -> State:
+        assert self.action_space.contains(action.arr)
+        dx, dy, magnetization = action.arr
+        if dx != 0.0 or dy != 0.0:
```
Comment (tomsilver) [inline src/envs/screws.py:203]: would anything be lost if we _always_ transition move, and then also transition (de)magnetize? if we ever try to learn options in this environment, the != 0.0 thing here is going to be an issue

File: src/envs/screws.py [hunk 31]
```diff
@@ -0,0 +1,482 @@
...
+        # Sample center locations for each of the screw clusters.
+        screw_cluster_xcenters = []
+        for _ in range(self._num_screw_clusters):
+            screw_cluster_xcenters.append(
+                rng.uniform(self.rz_x_lb,
+                            self.rz_x_ub - self._receptacle_width))
+
+        for _ in range(num_tasks):
+            screw_name_to_pos: Dict[str, Tuple[float, float]] = {}
+            num_screws = rng.choice(possible_num_screws)
```
Comment (tomsilver) [inline src/envs/screws.py:223]: should this be `rng.integers(1, possible_num_screws+1)`? (a) i don't think we'd want to allow for no screws and (b) we should be able to sample `possible_num_screws`, based on the name

**Comment by NishanthJKumar** [comment]
Really - that's actually quite unexpected; lemme look into this.

File: src/envs/screws.py [hunk 32]
```diff
@@ -0,0 +1,482 @@
...
+            return self._transition_magnetize(state)
+        return self._transition_demagnetize(state)
+
+    def _get_tasks(self, num_tasks: int, possible_num_screws: List[int],
+                   rng: np.random.Generator) -> List[Task]:
+        tasks = []
+
+        # Sample center locations for each of the screw clusters.
+        screw_cluster_xcenters = []
+        for _ in range(self._num_screw_clusters):
```
Comment (tomsilver) [inline src/envs/screws.py:216]: why are these clusters sampled once for all tasks? wouldn't it make more sense to resample for each task?

**Comment by NishanthJKumar** [comment]
confirmed i can replicate this behavior; investigating what the first skeleton is/why it's failing rn.

File: src/envs/screws.py [hunk 33]
```diff
@@ -0,0 +1,482 @@
...
+
+        # Create gripper object such that it is at the middle of
+        # the screen.
+        state_dict[self._robot] = {
+            "pose_x": (self.rz_x_lb + self.rz_x_ub) / 2.0,
+            "pose_y": (self.rz_y_lb + self.rz_y_ub) / 2.0
+        }
+
+        init_state = utils.create_state_from_dict(state_dict)
+        task = Task(init_state, goal_atoms)
```
Comment (tomsilver) [inline src/envs/screws.py:281]: add an assertion that `len(goal_atoms) == 1`

File: src/envs/screws.py [hunk 34]
```diff
@@ -0,0 +1,482 @@
...
+        return tasks
+
+    def _get_single_task(self, screw_name_to_pos: Dict[str, Tuple[float,
+                                                                  float]],
+                         rng: np.random.Generator) -> Task:
+        state_dict = {}
+        goal_atoms: Set[GroundAtom] = set()
+        # Select a random screw that needs to be in the receptacle as the goal
+        # GroundAtom.
+        goal_screw_name = rng.choice(list(screw_name_to_pos.keys()))
```
Comment (tomsilver) [inline src/envs/screws.py:248]: depending on how the dictionary is built, this may be nondeterministic. let's change the `list` to `sorted`

Comment (tomsilver) [reply src/envs/screws.py:248]: also, you can remove the `.keys()`. so:
`goal_screw_name = rng.choice(sorted(screw_name_to_pos))`

File: src/envs/screws.py [hunk 35]
```diff
@@ -0,0 +1,482 @@
...
+
+    def _surface_xy_is_valid(self, x: float, y: float,
+                             existing_xys: Set[Tuple[float, float]]) -> bool:
+        # First, check that there isn't another screw at this location.
+        for (existing_x, existing_y) in existing_xys:
+            if (abs(existing_x - x) < self._screw_width
+                    and abs(existing_y - y) < self._screw_height):
+                return False
+        # Next, check that the location is valid.
+        y_is_valid = y == self.rz_y_lb + self._screw_height / 2.0
```
Comment (tomsilver) [inline src/envs/screws.py:292]: can you add parentheses around the right hand sides here for easier reading?

File: src/envs/screws.py [hunk 36]
```diff
@@ -0,0 +1,482 @@
...
+                next_state.set(screw, "held", 0.0)
+
+        return next_state
+
+    def _GripperCanPickScrew_holds(self, state: State,
+                                   objects: Sequence[Object]) -> bool:
+        gripper, screw = objects
+        gripper_y = state.get(gripper, "pose_y")
+        screw_y = state.get(screw, "pose_y")
+        screw_maxy = screw_y + self._screw_height / 2.0
```
Comment (tomsilver) [inline src/envs/screws.py:374]: nit: add underscores after the 'max' and 'min' here and below

File: tests/envs/test_screws.py [hunk 37]
```diff
@@ -0,0 +1,72 @@
...
+        demagnetize_gripper_option
+    ])
+    traj = utils.run_policy_with_simulator(
+        policy,
+        env.simulate,
+        task.init,
+        lambda _: False,
+        max_num_steps=1000,
+        exceptions_to_break_on={utils.OptionExecutionFailure})
+    assert traj.states[-1].allclose(traj.states[-3])
```
Comment (tomsilver) [inline tests/envs/test_screws.py:72]: how does this assertion check what's described in the comment?

**Review by tomsilver — COMMENTED** [review]
looking good! I think we should be good after 1 more round

File: src/envs/screws.py (hunk 30 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:203]: yup - we can definitely do this and not lose anything! I'll go ahead and make this change

File: src/envs/screws.py (hunk 32 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:216]: good point! my bad; i missed this

File: tests/envs/test_screws.py (hunk 37 above)
Comment (NishanthJKumar) [reply tests/envs/test_screws.py:72]: added comment describing this

File: src/envs/screws.py (hunk 31 above)
Comment (NishanthJKumar) [reply src/envs/screws.py:223]: wait - `possible_num_screws` is a list that's passed in from `settings.py` (e.g. `[15, 20]`). I was intending this line to select a particular number from this list, which i believe is what it currently does

Comment (tomsilver) [reply src/envs/screws.py:223]: oh sorry, my bad, ignore!

**Comment by NishanthJKumar** [comment]
I ran `backchaining` and `cluster_and_intersect` strips learning just to verify. Results are that we solve `50/50` test tasks with backchaining and recover the ground truth operators. We solve `0/50` with cluster_and_intersect because planning always times out in grounding!

File: src/envs/screws.py [hunk 38]
```diff
@@ -0,0 +1,482 @@
...
+                     action: Optional[Action] = None,
+                     caption: Optional[str] = None) -> List[Image]:
+        fig, ax = plt.subplots(1, 1)
+        plt.xlim([
+            self.rz_x_lb - 2 * self._screw_width,
+            self.rz_x_ub + 2 * self._screw_width
+        ])
+        plt.ylim([self.rz_y_lb, self.rz_y_ub])
+
+        # Draw receptacle
```
Comment (tomsilver) [inline src/envs/screws.py:121]: nit: add periods after this and all comments

**Review by tomsilver — APPROVED** [review]
LGTM, awesome work!
